### PR TITLE
Add HIP/ROCm Support for AMD GPUs 

### PR DESCRIPTION
**Pull request by HPC-Ken**
## Description

This pull request is submitted on behalf of **AMD**, proposing support for **AMD GPUs via HIP/ROCm**. The contribution enables gsplat to use GPU acceleration on AMD hardware while preserving the existing CUDA implementation.

The approach is additive and non-intrusive: all original CUDA (`.cu`) kernels and CUDA build logic remain intact, while parallel HIP (`.hip`) implementations and build paths are introduced where appropriate. Two AMD GPU families are targeted with **separately tuned** kernels (RDNA / consumer and CDNA / datacenter); their per-variant code lives in dedicated `arch/radeon/` and `arch/instinct/` subtrees and is never unified with `#ifdef`s, so each backend retains its performance characteristics.

---

### 1. HIP / ROCm Build Support (Linux, Python / setup.py)

- New environment-variable selector at install time:
  ```bash
  export GSPLAT_BACKEND=cuda            # default — upstream CUDA path, unchanged
  export GSPLAT_BACKEND=rocm_radeon     # AMD RDNA  (gfx1100 / gfx1200)
  export GSPLAT_BACKEND=rocm_instinct   # AMD CDNA  (gfx942  / gfx950)
  pip install -e .
  ```

- `gsplat/cuda/build.py` was rewritten as a **tri-target dispatcher** that:
  - Reads `GSPLAT_BACKEND` and validates against `{cuda, rocm_radeon, rocm_instinct}`
  - Sets per-backend `PYTORCH_ROCM_ARCH` defaults (`gfx1100;gfx1200` for Radeon, `gfx942;gfx950` for Instinct)
  - Builds backend-specific source list, include path list, and compile flags
  - Calls the existing `torch.utils.cpp_extension.load(...)` JIT loader
  - Exports `get_build_parameters()` for `setup.py` to consume — `setup.py` itself remains **byte-identical to upstream**

- `gsplat/cuda/_backend.py` was extended (additive only) to detect ROCm via `torch.version.hip` and locate `hipcc` under `$ROCM_PATH/bin`, `$ROCM_HOME/bin`, `/opt/rocm/bin`, or `PATH`. The CUDA `_find_cuda_home()` fallback is preserved unchanged — same file works for both backends.

- CUDA and ROCm builds are mutually exclusive at install time and selected by `GSPLAT_BACKEND`. Omitting the variable preserves the upstream CUDA behavior exactly.

---

### 2. HIP Implementations of Existing CUDA Kernels

Twenty-two HIP source files were introduced as one-to-one ports of every upstream CUDA kernel, in two **separately tuned** variants:

- `arch/radeon/gsplat/cuda/csrc/*.hip` (22 files) — tuned for RDNA3 / RDNA4 (`gfx1100` / `gfx1200`), wave32, `-ffast-math`
- `arch/instinct/gsplat/cuda/csrc/*.hip` (22 files) — tuned for CDNA3 / CDNA4 (`gfx942` / `gfx950`), wave64, conservative math set (`-fno-math-errno`, `-ffp-contract=fast`, `-fdenormal-fp-math=preserve-sign`, `-fgpu-flush-denormals-to-zero`)

Per-pair divergence is small (4–124 lines per file) and concentrated in launch-bound macros, wave-size constants, and FAST_MATH knobs. The variants are never unified — `arch/<variant>/` subtrees are mutually exclusive at the source-list level.

Files that are byte-identical between Radeon and Instinct but differ from the upstream CUDA versions live in a shared subtree `arch/rocm/`:

- C++ dispatchers (Adam.cpp, Intersect.cpp, Lidar.cpp, Null.cpp, Projection.cpp, QuatScaleToCovar.cpp, Rasterization.cpp, Relocation.cpp, SphericalHarmonics.cpp)
- HIP-translated headers (Cameras.hip.h, ExternalDistortion.hip.h, Lidars.hip.h, Utils.hip.h, Projection2DGS.hip.h)
- ROCm-overlay headers (Cameras.h, CameraWrappers.h, ExternalDistortion.h, ExternalDistortionWrappers.h, Lidars.h, Ops.h, torch_rocm_compat.h)

A per-variant `Common.h` lives separately in `arch/radeon/gsplat/cuda/include/` and `arch/instinct/gsplat/cuda/include/` (wave size, launch-bound, and fast-math knobs differ between RDNA and CDNA, so it is not shared).

Source selection between `.cu` and `.hip` is performed at install time by `build.py` based on `GSPLAT_BACKEND`, with no runtime branching.

---

### 3. Shared HIP / ROCm Compatibility Headers

A small set of HIP-translation support headers was added under `third_party/`:

- `third_party/cgshim/cooperative_groups/reduce.h` — shared `cooperative_groups::reduce` polyfill matching CUDA's call signature
- `third_party/cgshim/cuda/std/{limits,optional,type_traits}` — `cuda::std::*` polyfills for HIP

Per-variant `cooperative_groups::reduce` shims live in `arch/<variant>/gsplat/cuda/include/hip_shim/cooperative_groups/reduce.h` (the only file that diverges between Radeon's wave32 and Instinct's wave64 reduction paths).

Additionally, `gsplat/cuda/csrc/third_party/glm/` was updated to **glm 1.0.3** and vendored in-tree (upstream tracks it as a git submodule). The newer release is required for clean HIP/clang compilation; the in-tree vendoring also removes the `git submodule update --init` step from the ROCm container build.

---

### 4. PyTorch Version Compatibility Shim

`arch/rocm/gsplat/cuda/include/torch_rocm_compat.h` re-introduces three PyTorch ROCm masquerade aliases that were dropped in the PyTorch 2.8 release but are present in 2.13+:

- `at::cuda::getCurrentCUDAStream` → forwarded to `c10::hip::getCurrentHIPStreamMasqueradingAsCUDA`
- `at::cuda::OptionalCUDAGuard` → aliased to `c10::hip::OptionalHIPGuardMasqueradingAsCUDA`
- `c10::cuda::CUDACachingAllocator::get` → forwarded to `c10::hip::HIPCachingAllocatorMasqueradingAsCUDA::get`

The shim is force-included on every ROCm translation unit via a `-include` flag added in `_rocm_flags()` in `build.py`, and is **version-guarded** (`TORCH_VERSION_MAJOR/MINOR < 2.13`) so it becomes a no-op on PyTorch 2.13+ where the aliases are already present. CUDA builds never include it.

---

### 5. Single Triple Target Dockerfile

A single `Dockerfile` at the repo root builds a self-contained image for any of the three backends, selected at build time:

```bash
docker build --build-arg GSPLAT_BACKEND=cuda          -t gsplat:cuda     .
docker build --build-arg GSPLAT_BACKEND=rocm_radeon   -t gsplat:radeon   .
docker build --build-arg GSPLAT_BACKEND=rocm_instinct -t gsplat:instinct .
```

The Dockerfile picks the correct base image per backend (`pytorch/pytorch:2.4.0-cuda12.4-cudnn9-devel` for CUDA; `rocm/pytorch:rocm7.2_ubuntu24.04_py3.12_pytorch_release_2.8.0` for ROCm), installs the right system + Python dependencies, and pre-warms the JIT extension so the first runtime `import gsplat` is instant. Build args allow overrides:

- `BASE_IMAGE_CUDA` / `BASE_IMAGE_ROCM` — point at a different base
- `PYTORCH_ROCM_ARCH` — narrow the AMD arch list (e.g. just `gfx1100`)
- `TORCH_CUDA_ARCH_LIST` — narrow the NVIDIA arch list (default covers Turing → Hopper)

A backend-aware install script (`docker/install_examples_deps.sh`) handles the example-script Python dependencies:

- **CUDA**: full upstream `examples/requirements.txt` minus `ppisp` (NVIDIA ISP, brittle nvcc build); `ppisp` is reattempted as non-fatal optional
- **ROCm**: ROCm-safe subset (drops `nvidia-ncore`, `ppisp`, `nerfacc` — none have ROCm wheels), then a *patched* build of `fused-bilagrid` (uint64-mask fix for ROCm 7.x wave64 + HIP-aware `setup.py` pointing at the per-variant `hip_shim/` include path), plus the PyTorch 2.8 `cpp_extension.py` hipify-None bug-fix sed patch

The Dockerfile requires BuildKit (modern Docker default; for older daemons `export DOCKER_BUILDKIT=1`). A `.dockerignore` excludes scratch dirs, datasets, and training output from the build context.

---

### 6. README and Documentation Updates

The main `README.md` was updated to include:

- A `## Building on ROCm` section describing the tri-target backend selection
- A `### Selecting a backend` table listing values, toolchains, default architectures
- Per-backend `### Build` sections (Radeon, Instinct, CUDA-unchanged)
- A `### Running with Docker` section with the three `docker build` invocations
- A `### Running the tests on ROCm` subsection with per-backend `pytest` commands and known per-variant test caveats

One new documentation file was added:

- `arch/README.md` — explains the `arch/{radeon,instinct,rocm}/` layout for users browsing the source

---

### 7. Additional Robustness Improvements

- `setup.py` is preserved byte-identical to upstream, but uses the new `gsplat/cuda/build.py` automatically via its `importlib.util.spec_from_file_location("gsplat_cuda_build", ...)` dynamic load
- `pyproject.toml`, `LICENSE`, `CITATION.bib`, `pytest.ini`, `formatter.sh`, `EXPLORATION.md`, every `.cu`/`.cuh`/`.cpp`/`.h` under `gsplat/cuda/csrc/` and `gsplat/cuda/include/`, every Python module under `gsplat/` (except `_backend.py` and `build.py`), every test, every example, every doc, every asset is byte-identical to upstream. CUDA users see no change.
- The merger applied no `#ifdef`-based unification of the two AMD variants; per-variant code is fully separated in `arch/radeon/` and `arch/instinct/` subtrees
- All upstream `.cu` files are preserved at their original paths (the CUDA backend depends on them)

---

## Tested Configurations

All changes were validated internally at AMD using **ROCm 7.2** on the following platforms:

- **RDNA3 GPU (Radeon RX 7900 XTX, `gfx1100`)**
  - Container: `rocm/pytorch:rocm7.2_ubuntu24.04_py3.12_pytorch_release_2.8.0`
  - Built via `docker build --build-arg GSPLAT_BACKEND=rocm_radeon -t gsplat:radeon .` → 32 GB image
  - Mip-NeRF 360 `garden` scene, `simple_trainer.py default --data_factor 4`: 27 it/s steady-state on AMD Radeon RX 7900 XTX, training loss converging normally; 4500-step partial run validated with no kernel errors
  - Upstream `pytest tests/`: 44 passed, 7 environmental failures (missing CUDA-only optional Python deps `nerfacc` and `PLAS`), 33 skipped — same skip set as upstream

- **CDNA4 GPU (AMD Instinct MI350X, `gfx950`)**
  - Container: same release ROCm base
  - Built via `docker build --build-arg GSPLAT_BACKEND=rocm_instinct -t gsplat:instinct .` → 32 GB image
  - JIT compile of all 32 source files for `gfx950`: clean
  - Upstream `pytest tests/`: 44 passed / 7 env-fail / 33 skipped (identical to Radeon, on the same release base)

- **NVIDIA Turing GPU (Quadro RTX 6000, `sm_75`) — regression check**
  - Container: `pytorch/pytorch:2.4.0-cuda12.4-cudnn9-devel`
  - Built via `docker build --build-arg GSPLAT_BACKEND=cuda -t gsplat:cuda .` → 16 GB image
  - Same `garden` training: 30 000 / 30 000 steps, final loss 0.026, 13.8 it/s
  - Upstream `pytest tests/`: 50 passed / 1 env-fail (PLAS missing) / 33 skipped — **byte-for-byte identical** to the pristine upstream test result on the same hardware
  - Confirms the tri-target dispatcher in `build.py` does not regress the CUDA path

No functional regressions were observed during testing on any of the three backends.

---

## Feedback and Next Steps

Feedback on structure, scope, and implementation details is welcome. Adjustments can be made to align more closely with project conventions or maintainer preferences.

— Ken O'Brien,  Andrea Fondacaro

### CODE REVIEW

File: arch/rocm/gsplat/cuda/include/Ops.h [hunk 1]
```diff
@@ -0,0 +1,680 @@
...
+    bool use_hit_distance,
+    // forward outputs
+    const at::Tensor &render_alphas, // [..., C, image_height, image_width, 1]
+    const at::Tensor &last_ids,      // [..., C, image_height, image_width]
+    // gradients of outputs
+    const at::Tensor &v_render_colors, // [..., C, image_height, image_width, 3]
+    const at::Tensor &v_render_alphas, // [..., C, image_height, image_width, 1]
+    const at::optional<at::Tensor> &v_render_normals // [..., C, image_height, image_width, 3]
+);
+} // namespace gsplat
```
Comment (jungx098) [inline arch/rocm/gsplat/cuda/include/Ops.h:680]: This ROCm `Ops.h` overlay is missing the `mcmc_perturb_positions` declaration. `ext.cpp` (included in the ROCm source list via `_rocm_sources()`) binds `&gsplat::mcmc_perturb_positions`, so without the decl here the ROCm build fails to compile:

```
gsplat/hip/ext.cpp:828: error: 'mcmc_perturb_positions' is not a member of 'gsplat'
```

30/31 registered ops are covered in the overlay — only this one is missing, and it fails on the PyTorch 2.8 stack too. Fix: declare it here, and add the dispatcher (`arch/rocm/.../csrc/MCMCPerturb.cpp`) + kernel (`arch/{radeon,instinct}/.../csrc/MCMCPerturbHIP.hip`).

_Tested: commit `826c31a`, `GSPLAT_BACKEND=rocm_radeon` (gfx1150), PyTorch `2.12.0+rocm7.1` / ROCm `7.1.52802`, AMD Radeon 890M, Ubuntu (Linux 6.17)._

File: arch/rocm/gsplat/cuda/include/torch_rocm_compat.h [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+#if defined(USE_ROCM) || defined(__HIP_PLATFORM_AMD__)
+
+#include <torch/csrc/api/include/torch/version.h>
+
+// Apply only when the alias macros aren't already provided by PyTorch.
+// PyTorch 2.13+ ships them; 2.8.x does not.
+#if !defined(TORCH_VERSION_MAJOR) || \
+    (TORCH_VERSION_MAJOR < 2) || \
+    (TORCH_VERSION_MAJOR == 2 && TORCH_VERSION_MINOR < 13)
```
Comment (jungx098) [inline arch/rocm/gsplat/cuda/include/torch_rocm_compat.h:41]: The `< 2.13` guard is wrong for 2.9–2.12: `torch==2.12.0+rocm7.1` already ships these aliases, so the shim redeclares them and every ROCm TU fails to compile:

```
:56 error: no viable conversion 'c10::cuda::CUDAStream' -> 'c10::hip::HIPStreamMasqueradingAsCUDA'
:72 error: ambiguating new declaration of 'c10::cuda::CUDACachingAllocator::get()'
```

Bumping to `< 2.12` just moves the problem to the next wheel. Recommend a configure-time feature probe (e.g. pass `-DGSPLAT_HAS_CUDA_MASQUERADE` when `at::cuda::getCurrentCUDAStream` compiles) instead of a version gate.

_Tested: commit `826c31a`, `GSPLAT_BACKEND=rocm_radeon` (gfx1150), PyTorch `2.12.0+rocm7.1` / ROCm `7.1.52802`, AMD Radeon 890M, Ubuntu (Linux 6.17)._

**Review by charyang-ai — COMMENTED** [review]
any performance data on Radeon card?

**Comment by HPC-Ken** [comment]
@charyang-ai Will answer internally.

**Comment by adamdbrw** [comment]
@HPC-Ken I would like to run with this PR in, what are the current gaps?

**Comment by HPC-Ken** [comment]
Hi @adamdbrw, I'll mail you. We're working on merging a few efforts.

**Comment by HPC-Ken** [comment]
Hi @eust-w. For ROCm support, if you're merging kernels, please use the ones from this PR, as they are the most optimized.
